### PR TITLE
Add Product Config Endpoints

### PR DESCRIPTION
**Pull request by luispuentesvega**
# PR Details

- Add Product Config Endpoints

### ERD Diagram
DB drawio

## Description

- Implement Singleton parent for the DB Connection
- Add Product Category List endpoint
- Add Product Category Create endpoint

### CODE REVIEW

**Comment by fabiancnieto** [comment]
let's see

**Review by fabiancnieto — COMMENTED** [review]
another try

File: .gitignore [hunk 1]
```diff
@@ -4,4 +4,6 @@ package-lock.json
 *.log
 *.gz
 dist
-.env
\ No newline at end of file
+.env
+.vscode
+dist/*
```
Comment (fabiancnieto) [inline .gitignore:9]:
```suggestion
dist/*

```

File: src/index.ts [hunk 2]
```diff
@@ -1,8 +1,30 @@
+import 'reflect-metadata';
+import { container } from './inversify.config';
+
```
Comment (fabiancnieto) [inline src/index.ts:3]:
```suggestion
import { container } from './inversify.config';
```

File: src/config/DynamoClient.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
+import { IDbClient } from "./interfaces";
+import { injectable } from "inversify";
+import { DynamoDB } from "aws-sdk";
+
+@injectable()
+export class DynamoClient implements IDbClient {
+  private _instance: any;
+
+  constructor() {
```
Comment (fabiancnieto) [inline src/config/DynamoClient.ts:9]:
```suggestion
  constructor(dynamoDbClient: DynamoDB) {
```

File: src/controllers/ProductCategoryController.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import { injectable, inject } from "inversify";
+
+import { Request, Response } from "express";
+import TYPES from "../config/types";
+import { IProductCategoryController, IProductCategoryService } from "../config/interfaces";
+
+@injectable()
+export default class ProductCategoryController implements IProductCategoryController {
+  @inject(TYPES.IProductCategoryService) private _productCategoryService: IProductCategoryService;
```
Comment (fabiancnieto) [inline src/controllers/ProductCategoryController.ts:9]:
```suggestion
  constructor(
      @inject(TYPES.IProductCategoryService) private readonly productCategoryService: IProductCategoryService
  ) {}
```

File: src/controllers/ProductCategoryController.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+  async getAllProductCategories(req: Request, res: Response) {
```
Comment (fabiancnieto) [inline src/controllers/ProductCategoryController.ts:11]:
```suggestion
  public async getAllProductCategories(req: Request, res: Response) {
```

File: tsconfig.json [hunk 6]
```diff
@@ -1,16 +1,19 @@
 {
   "compilerOptions": {
-    "module": "commonjs",
+    "module": "commonjs", // how to import and export
```
[the remark this answers is not in the document]
Comment (luispuentesvega) [reply tsconfig.json:3]: Added!

File: package.json [hunk 7]
```diff
@@ -6,12 +6,19 @@
...
     "start": "tsc & node .",
-    "dev": "tsc -w & nodemon ."
+    "dev": "tsc -w & nodemon .",
+    "test": "jest"
   },
   "author": "luispuentesvega@gmail.com",
   "license": "ISC",
   "dependencies": {
-    "express": "^4.17.1"
+    "@types/jest": "^27.0.2",
```
[the remark this answers is not in the document]
Comment (luispuentesvega) [reply package.json]: Moved as a dev dependency

File: .gitignore (hunk 1 above)
Comment (luispuentesvega) [reply .gitignore]: done

File: src/controllers/ProductCategoryController.ts [hunk 8]
```diff
@@ -0,0 +1,22 @@
+import { injectable, inject } from "inversify";
+
+import { Request, Response } from "express";
```
Comment (fabiancnieto) [inline src/controllers/ProductCategoryController.ts:3]:
```suggestion
import { injectable, inject } from "inversify";
import { Request, Response } from "express";
```

**Review by fabiancnieto — APPROVED** [review]
please fix some minor comments and let's merge

File: jest.config.js [hunk 9]
```diff
@@ -0,0 +1,7 @@
+/** @type {import('ts-jest/dist/types').InitialOptionsTsJest} */
```
[the remark this answers is not in the document]
Comment (luispuentesvega) [reply jest.config.js:1]: It's an annotation to import the initial TS config for jest. It is added by default when starting jest in the project

Screen Shot 2021-12-19 at 3 22 17 PM

File: src/controllers/ProductCategoryController.ts (hunk 4 above)
Comment (luispuentesvega) [reply src/controllers/ProductCategoryController.ts]: Done!

File: .gitignore [hunk 10]
```diff
@@ -4,4 +4,7 @@ package-lock.json
...
+coverage/*
```
[the remark this answers is not in the document]
Comment (luispuentesvega) [reply .gitignore]: Done!

File: src/config/DynamoClient.ts [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+  }
+
+  public async addData<T>(tableName: string, item: T): Promise<void> {
+    const params = {
+      TableName: tableName,
+      Item: item
+    };
+    this.instance.put(params).promise();
+  }
+}
```
[the remark this answers is not in the document]
Comment (luispuentesvega) [reply src/config/DynamoClient.ts]: Done!

File: src/config/types.ts [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+
+export type QueryData<T> = {
+  data: Array<T>;
+  count: number;
+};
+
+export type DbResult<T> = {
+  Items: Array<T> | [];
+  Count: number;
+  ScannedCount: number;
```
[the remark this answers is not in the document]
Comment (luispuentesvega) [reply src/config/types.ts]: Done!

File: src/controllers/ProductCategoryController.ts (hunk 8 above)
Comment (luispuentesvega) [reply src/controllers/ProductCategoryController.ts]: Done!

**Comment by luispuentesvega** [comment]
@fabiancnieto . Fixed the minor comments. Thank you!

**Pull request merged**